sidechain/consensus/slots/src/slots.rs

```diff
@@ -73,7 +73,7 @@ impl<ParentchainBlock: ParentchainBlockTrait> SlotInfo<ParentchainBlock> {
 			slot,
 			timestamp,
 			duration,
-			ends_at: timestamp + time_until_next_slot(duration),
+			ends_at: duration_now() + time_until_next_slot(duration),
```

Comment:
This seems to have been the problem. `time_until_next_slot` calculates the remaining time from `now` but we have added it to a `timestamp` which has been set some time ago. This has caused some inaccuracies of the slot time ending

Reply:
I'm not sure why the `ends_at` calculation would invalidate the slot timestamp? It simply calculates the time left of the current time slot, the start time does not matter. But the `ends_at` calculation is surely not very intuitive here, but that's because the `time_until_next_slot` was most likely designed for the `SlotStream` 
https://github.com/integritee-network/worker/blob/83db66d655a6fcff0e3416c15c343762b004a159/sidechain/consensus/slots/src/slot_stream.rs#L61
and is reused here. But I don't think we should reinvent something that's already in use: https://github.com/paritytech/substrate/blob/master/client/consensus/slots/src/slots.rs#L88

Reply:
Oh wow, detective Haerdi! Cool, but I am still kind of surprised why this is an issue.

Reply:
I don't understand it completely as well, but my best guess is: 
There has been a significant time drift of `ends_at`, and there for some too long `remaining times`:
https://github.com/integritee-network/worker/blob/5b5ed0d4c67a4b8598548d3ea2288bacc4dea6d7/sidechain/consensus/slots/src/lib.rs#L301
and 
https://github.com/integritee-network/worker/blob/5b5ed0d4c67a4b8598548d3ea2288bacc4dea6d7/sidechain/consensus/slots/src/lib.rs#L184

and a misjudgement of 
https://github.com/integritee-network/worker/blob/5b5ed0d4c67a4b8598548d3ea2288bacc4dea6d7/sidechain/consensus/slots/src/slots.rs#L82-L93

In corner cases, this lead to successfully proposing a block with a timestamp outside the range of the actual claimed slot. Hence, the block will be gossiped and reimported by the workers. And there the `Slot` will be calculated using the block `timestamp` and the const `slot_duration` https://github.com/integritee-network/worker/blob/5b5ed0d4c67a4b8598548d3ea2288bacc4dea6d7/sidechain/consensus/slots/src/slots.rs#L95-L97
which will end up in a different `Slot` then the proposing worker thought the block is in. Leading to the `InvalidAuthority`

Does that sound plausible?

Reply:
Thank you very much for this elaborate explanation. However, as I understand the code, your change would actually introduce the potential drift of `ends_at` and before by using the already defined `timestamp`, we ensured that there is no drift.

I also wonder because proposing remaining duration has a safety parameter, which should mitigate some corner cases:

https://github.com/integritee-network/worker/blob/5b5ed0d4c67a4b8598548d3ea2288bacc4dea6d7/sidechain/consensus/aura/src/lib.rs#L228

However, if you are certain that by your change you achieve:

* no longer propose invalid blocks

instead of:

* accepting invalid blocks

I will double-check the code, if I can come to this conclusion.

Reply:
That was my first impression as well, but the `time_until_next_slot` does not work the way you would expect. @haerdib wrote a test (further down) to demonstrate we get wrong results before this change

Reply:
Wow, this is so intuitive, but I am convinced. Thanks @haerdib for the test! :D

Reply:
At least about "accepting invalid blocks" you don't have to worry, as the `ends_at` is only used in the block production. Verifying the block author is completely independent on the `SlotInfo` and therefore the changes `ends_at` parameter.

Reply:
Yes, you are right! I can see that know, thanks!